File: ios-catalogue/App/Components/CheckboxShowcase.swift
import SwiftUI
import CoreUI
import Components

struct CheckboxShowcase: View {
    @State private var checked: Bool = false
    @State private var label: String = "Email me product updates"
    @State private var helper: String = "Roughly twice a month. No spam."
    @State private var enabled: Bool = true
    @State private var error: Bool = false

    var body: some View {
        PlaygroundScaffold(
            preview: {
                PrismaCheckbox(
                    checked: $checked,
                    label: label.isEmpty ? nil : label,
                    helperText: helper.isEmpty ? nil : helper,
                    enabled: enabled,
                    isError: error
                )
            },
            knobs: {
                BoolKnobRow(label: "Checked", value: $checked)
                BoolKnobRow(label: "Enabled", value: $enabled)
                BoolKnobRow(label: "Error", value: $error)
                StringKnobRow(label: "Label", value: $label)
                StringKnobRow(label: "Helper", value: $helper)
            },
            states: {
                StateCell("Unchecked") { PrismaCheckbox(checked: .constant(false), label: "Default unchecked") }
                StateCell("Checked") { PrismaCheckbox(checked: .constant(true), label: "Default checked") }
                StateCell("Indeterminate") { PrismaCheckbox(state: .constant(.indeterminate), label: "Indeterminate") }
                StateCell("Disabled checked") { PrismaCheckbox(checked: .constant(true), label: "Locked on", enabled: false) }
                StateCell("Disabled unchecked") { PrismaCheckbox(checked: .constant(false), label: "Locked off", enabled: false) }
                StateCell("Error") { PrismaCheckbox(checked: .constant(false), label: "Required", helperText: "This field is required.", isError: true) }
                StateCell("Group + parent") { GroupExample() }
            },
            code: {
                var lines: [String] = ["PrismaCheckbox("]
                lines.append("    checked: $checked,")
                if !label.isEmpty { lines.append("    label: \"\(label)\",") }
                if !helper.isEmpty { lines.append("    helperText: \"\(helper)\",") }
                if !enabled { lines.append("    enabled: false,") }
                if error { lines.append("    isError: true,") }
                lines.append(")")
                return lines.joined(separator: "\n")
            },
            a11y: {
                A11yPanel(
                    role: ".isToggle (tri-state via accessibilityValue)",
                    minTouchTarget: "44 × 44 pt",
                    bullets: [
                        "State (checked / unchecked / indeterminate) announced; label read in same pass.",
                        "Indeterminate is for parent groups — children remain individual booleans.",
                        "Error appends \"required\" or the error message via accessibilityHint.",
                        "Disabled communicated via the role; visual dim is supporting, not primary."
                    ]
                )
            }
        )
    }
}

private struct GroupExample: View {
    @State private var alpha: Bool = true
    @State private var beta: Bool = false
    @State private var gamma: Bool = true

    private var parent: PrismaCheckboxState {
        if alpha && beta && gamma { return .checked }
        if !alpha && !beta && !gamma { return .unchecked }
        return .indeterminate
    }

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp1) {
            PrismaCheckbox(
                state: Binding(
                    get: { parent },
                    set: { _ in
                        let n = parent != .checked
                        alpha = n; beta = n; gamma = n
                    }
                ),
                label: "Notifications"
            )
            VStack(alignment: .leading, spacing: PrismaSpacing.sp1) {
                PrismaCheckbox(checked: $alpha, label: "Mentions")
                PrismaCheckbox(checked: $beta, label: "DMs")
                PrismaCheckbox(checked: $gamma, label: "Releases")
            }
            .padding(.leading, PrismaSpacing.sp7)
        }
    }
}

File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/components/CheckboxShowcase.kt
package xyz.ksharma.prisma.catalogue.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.state.ToggleableState
import xyz.ksharma.prisma.catalogue.playground.A11yPanel
import xyz.ksharma.prisma.catalogue.playground.BoolKnobRow
import xyz.ksharma.prisma.catalogue.playground.PlaygroundScaffold
import xyz.ksharma.prisma.catalogue.playground.StateCell
import xyz.ksharma.prisma.catalogue.playground.StringKnobRow
import xyz.ksharma.prisma.components.checkbox.PrismaCheckbox
import xyz.ksharma.prisma.components.checkbox.PrismaTriStateCheckbox
import xyz.ksharma.prisma.tokens.PrismaSpacing

@Composable
public fun CheckboxShowcase() {
    var checked by rememberSaveable { mutableStateOf(false) }
    var label by rememberSaveable { mutableStateOf("Email me product updates") }
    var helper by rememberSaveable { mutableStateOf("Roughly twice a month. No spam.") }
    var enabled by rememberSaveable { mutableStateOf(true) }
    var error by rememberSaveable { mutableStateOf(false) }

    PlaygroundScaffold(
        preview = {
            PrismaCheckbox(
                checked = checked,
                onCheckedChange = if (enabled) ({ checked = it }) else null,
                label = label.takeIf { it.isNotBlank() },
                helperText = helper.takeIf { it.isNotBlank() },
                enabled = enabled,
                isError = error,
            )
        },
        knobs = {
            BoolKnobRow("Checked", checked, { checked = it })
            BoolKnobRow("Enabled", enabled, { enabled = it })
            BoolKnobRow("Error", error, { error = it })
            StringKnobRow("Label", label, { label = it })
            StringKnobRow("Helper", helper, { helper = it })
        },
        code = {
            buildString {
                append("PrismaCheckbox(\n")
                append("    checked = checked,\n    onCheckedChange = { checked = it },\n")
                if (label.isNotBlank()) append("    label = \"${label}\",\n")
                if (helper.isNotBlank()) append("    helperText = \"${helper}\",\n")
                if (!enabled) append("    enabled = false,\n")
                if (error) append("    isError = true,\n")
                append(")")
            }
        },
        states = {
            StateCell("Unchecked") { PrismaCheckbox(checked = false, onCheckedChange = {}, label = "Default unchecked") }
            StateCell("Checked") { PrismaCheckbox(checked = true, onCheckedChange = {}, label = "Default checked") }
            StateCell("Indeterminate") { PrismaTriStateCheckbox(state = ToggleableState.Indeterminate, onClick = {}, label = "Indeterminate") }
            StateCell("Disabled checked") { PrismaCheckbox(checked = true, onCheckedChange = null, label = "Locked on", enabled = false) }
            StateCell("Disabled unchecked") { PrismaCheckbox(checked = false, onCheckedChange = null, label = "Locked off", enabled = false) }
            StateCell("Error") {
                PrismaCheckbox(
                    checked = false,
                    onCheckedChange = {},
                    label = "Required",
                    helperText = "This field is required.",
                    isError = true,
                )
            }
            StateCell("Group + parent") { GroupExample() }
        },
        a11y = {
            A11yPanel(
                role = "Checkbox / TriStateCheckbox",
                minTouchTarget = "48 × 48 dp",
                bullets = listOf(
                    "State (checked / unchecked / indeterminate) is announced; the label is read in the same pass.",
                    "Indeterminate is for parent groups — children remain individual booleans.",
                    "Error state appends \"required\" or the error message via stateDescription.",
                    "Disabled is conveyed via the role; the visual dimming alone is not sufficient.",
                ),
            )
        },
    )
}

@Composable
private fun GroupExample() {
    var alpha by rememberSaveable { mutableStateOf(true) }
    var beta by rememberSaveable { mutableStateOf(false) }
    var gamma by rememberSaveable { mutableStateOf(true) }

    val parent = when {
        alpha && beta && gamma -> ToggleableState.On
        !alpha && !beta && !gamma -> ToggleableState.Off
        else -> ToggleableState.Indeterminate
    }

    Column(verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp1)) {
        PrismaTriStateCheckbox(
            state = parent,
            onClick = {
                val newValue = parent != ToggleableState.On
                alpha = newValue; beta = newValue; gamma = newValue
            },
            label = "Notifications",
        )
        Column(modifier = Modifier.padding(start = PrismaSpacing.Sp7)) {
            PrismaCheckbox(checked = alpha, onCheckedChange = { alpha = it }, label = "Mentions")
            PrismaCheckbox(checked = beta, onCheckedChange = { beta = it }, label = "DMs")
            PrismaCheckbox(checked = gamma, onCheckedChange = { gamma = it }, label = "Releases")
        }
    }
}
